File: 2019/rust/src/bin/day25.rs
use std::cmp::Ordering;
use std::collections::HashSet;
use std::convert::TryInto;
use std::fmt::{Debug, Display};
use std::mem;

use num_traits::{Num, ToPrimitive};

use aoc2019::intcode::*;
use aoc2019::sparse::*;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
enum Door {
    North,
    East,
    South,
    West,
}
fn rev(s: &Door) -> Door {
    match s {
        Door::North => Door::South,
        Door::East => Door::West,
        Door::South => Door::North,
        Door::West => Door::East,
    }
}

impl Display for Door {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Door::North => write!(fmt, "north"),
            Door::East => write!(fmt, "east"),
            Door::South => write!(fmt, "south"),
            Door::West => write!(fmt, "west"),
        }
    }
}

impl std::str::FromStr for Door {
    type Err = String;
    fn from_str(s: &str) -> Result<Door, String> {
        match s {
            "north" => Ok(Door::North),
            "east" => Ok(Door::East),
            "south" => Ok(Door::South),
            "west" => Ok(Door::West),
            _ => Err(s.to_string()),
        }
    }
}

#[derive(Debug)]
struct Room {
    name: String,
    desc: String,
    doors: Vec<Door>,
    items: Vec<String>,
}

#[must_use]
fn ensure(b: bool) -> Option<()> {
    if b {
        Some(())
    } else {
        None
    }
}

enum ParsedRoom {
    Room(Room),
    EjectedFromCheckpoint(Room), // we found the checkpoint, but were ejected back to this room
}
fn parse_room(s: &str) -> Option<ParsedRoom> {
    let mut ls = s.lines();
    ensure("" == ls.next()?)?;
    ensure("" == ls.next()?)?;
    ensure("" == ls.next()?)?;
    let title = ls.next().unwrap();
    let name = title.trim_matches('=').trim().to_string();
    // The desc may be overwritten when we get through the security door,
    // as the desc there is "Analysing...", and the real description comes later
    let mut desc = ls.next().unwrap().to_string();
    assert_eq!(Some(""), ls.next());
    assert_eq!(Some("Doors here lead:"), ls.next());
    let mut doors = Vec::new();
    while let Some(l) = ls.next() {
        if l.starts_with("- ") {
            doors.push(l.split_at(2).1.parse().unwrap());
        } else {
            assert_eq!("", l);
            break;
        }
    }
    let mut items = Vec::new();
    match ls.next() {
        Some("Items here:") => {
            while let Some(l) = ls.next() {
                if l.starts_with("- ") {
                    items.push(l.split_at(2).1.to_string());
                } else {
                    assert_eq!("", l);
                    assert_eq!(Some("Command?"), ls.next());
                    break;
                }
            }
        }
        Some("Command?") => {}
        Some(l) => {
            if l.starts_with("A loud, robotic voice says \"Alert! Droids on this ship are") {
                // found security door
                let ejected : String = ls.flat_map(|e| vec![e,"\n"]).collect();
                match parse_room(&ejected)? {
                    ParsedRoom::Room(room) => return Some(ParsedRoom::EjectedFromCheckpoint(room)),
                    ParsedRoom::EjectedFromCheckpoint(_) => panic!("Huh, two checkpoints?"),
                }
            } else if l.starts_with("A loud, robotic voice says \"Analysis complete! You may proceed.\" and you enter the cockpit.") {
            // Got through the part a security door
                desc = l.to_string();
                while let Some(l) = ls.next() {
                    desc.push_str("\n");
                    desc.push_str(l);
                }
            } else {
                panic!("Unexpected line in room: {}",l)
            }
        }
        None => panic!("Unexpected end of room"),
    }
    assert_eq!(None, ls.next());
    Some(ParsedRoom::Room(Room {
        name,
        desc,
        doors,
        items,
    }))
}
struct Game<T, C: Vector<T>> {
    mach: Option<Box<dyn FnOnce(&str) -> IC<T, C>>>, // None: has halted
    room: Room,
    inv: Vec<String>,
    last_pickup: Option<String>, // some items are traps, remember what we last picked up in case it was a bad idea
}
#[derive(Debug)]
enum Cmd {
    Move(Door),
    Take(String),
    Drop(String),
}

impl<T, C: Vector<T>> Game<T, C> {
    // Will return last item picked up if we realise it was a trap
    // in that case, self may be in an inconsistent state.
    // In the case we move into the checkpoint, and get ejected, we return Ok(true)
    fn do_action(&mut self, cmd: Cmd) -> Result<bool, String>
    where
        C: 'static,
        T: 'static,
        T: Clone,
        T: TryInto<usize>,
        <T as TryInto<usize>>::Error: Debug,
        T: ToPrimitive,
        T: From<u8>,
        T: Num,
        T: PartialOrd,
        T: Display,
    {
        fn untrap<X>(item: Option<String>, x: Option<X>) -> Result<X, String> {
            match x {
                None => return Err(item.expect("nothing picked up, but found trap?")),
                Some(y) => return Ok(y),
            }
        };
        let mach = mem::replace(&mut self.mach, None); //This is a hack to avoid rust failing to move self.mach, even though we replace it later.
        let mach = mach.expect("The game is over");
        let new_mach;
        let mut ejected = false;
        match cmd {
            Cmd::Move(d) => {
                let ds = d.to_string();
                let (rm, m) = run_to_in(mach(&ds));
                new_mach = m;
                let room = untrap(self.last_pickup.clone(), parse_room(&rm))?;
                match room {
                    ParsedRoom::Room(r) => self.room = r,
                    ParsedRoom::EjectedFromCheckpoint(r) => {
                        ejected = true;
                        self.room = r
                    }
                }
            }
            Cmd::Take(i) => {
                let cmd = format!("take {}", i);
                let expected = format!("\nYou take the {}.\n\nCommand?\n", i);
                self.last_pickup = Some(i.clone());
                let (_, m) = untrap(
                    self.last_pickup.clone(),
                    run_to_in_expect(&expected, mach(&cmd)),
                )?;
                new_mach = m;
                let j = self
                    .room
                    .items
                    .iter()
                    .position(|item| *item == i)
                    .expect("bad item");
                self.inv.push(i);
                self.room.items.remove(j);
            }
            Cmd::Drop(i) => {
                let cmd = format!("drop {}", i);
                let expected = format!("\nYou drop the {}.\n\nCommand?\n", i);
                let (_, m) = untrap(
                    self.last_pickup.clone(),
                    run_to_in_expect(&expected, mach(&cmd)),
                )?;
                new_mach = m;
                let j = self
                    .inv
                    .iter()
                    .position(|item| *item == i)
                    .expect("bad item");
                self.room.items.push(i);
                self.inv.remove(j);
            }
        }
        self.mach = new_mach;
        Ok(ejected)
    }
}

fn run_to_in_expect<T: 'static, C: 'static>(
    expect: &str,
    mach: IC<T, C>,
) -> Option<(String, Option<Box<dyn FnOnce(&str) -> IC<T, C>>>)>
// TODO: I really should abstract this list of requirements!
where
    C: Vector<T>,
    T: Clone,
    T: TryInto<usize>,
    <T as TryInto<usize>>::Error: Debug,
    T: ToPrimitive,
    T: From<u8>,
    T: Num,
    T: PartialOrd,
    T: Display,
{
    run_to_in_worker(Some(expect), mach)
}

fn run_to_in<T: 'static, C: 'static>(
    mach: IC<T, C>,
) -> (String, Option<Box<dyn FnOnce(&str) -> IC<T, C>>>)
where
    C: Vector<T>,
    T: Clone,
    T: TryInto<usize>,
    <T as TryInto<usize>>::Error: Debug,
    T: ToPrimitive,
    T: From<u8>,
    T: Num,
    T: PartialOrd,
    T: Display,
{
    run_to_in_worker(None, mach).unwrap()
}

// Runs the machine just enough to check if the output is as we expect (if anything),
// then either returns None, or continues running until we want input or halt.
// Thus, if expect = None, the return will always be Some(...)
fn run_to_in_worker<T: 'static, C: 'static>(
    expect: Option<&str>,
    mut mach: IC<T, C>,
) -> Option<(String, Option<Box<dyn FnOnce(&str) -> IC<T, C>>>)>
where
    C: Vector<T>,
    T: Clone,
    T: TryInto<usize>,
    <T as TryInto<usize>>::Error: Debug,
    T: ToPrimitive,
    T: From<u8>,
    T: Num,
    T: PartialOrd,
    T: Display,
{
    let e_len = expect.map(str::len);
    let mut got_expected = expect.is_none();
    let mut out = String::new();
    loop {
        match mach.run_to_io() {
            IO::Out(m, c) => {
                let ch: char = c.to_u8().unwrap().into();
                out.push(ch);
                match e_len.map(|l| l.cmp(&out.len())) {
                    Some(Ordering::Equal) => {
                        ensure(expect.unwrap() == out)?;
                        got_expected = true;
                    }
                    Some(Ordering::Less) => return None,
                    _ => {}
                }
                mach = m;
            }
            IO::In(f) => {
                ensure(got_expected)?;
                return Some((out, Some(Box::new(|s| give_string(f, &s)))));
            }
            IO::Halt => {
                ensure(got_expected)?;
                return Some((out, None));
            }
        }
    }
}

// This function adds a newline
fn give_string<T: 'static, C: 'static>(
    mut mach: Box<dyn FnOnce(T) -> IC<T, C>>,
    s: &str,
) -> IC<T, C>
where
    C: Vector<T>,
    T: Clone,
    T: TryInto<usize>,
    <T as TryInto<usize>>::Error: Debug,
    T: ToPrimitive,
    T: From<u8>,
    T: Num,
    T: PartialOrd,
    T: Display,
{
    let mut chars = s.chars();
    while let Some(c) = chars.next() {
        let t = c as u32;
        let t: u8 = t.try_into().unwrap();
        let t: T = t.into();
        match mach(t).run_to_io() {
            IO::In(f) => mach = f,
            _ => panic!("expected machine to want input"),
        }
    }
    let newline = '\n' as u32;
    let newline: u8 = newline.try_into().unwrap();
    let newline: T = newline.into();
    mach(newline)
}

// PART A, General Strategy
// It appears the ship is a tree (with back-edges), so:
// do a pre-order traversal, picking up all items and finding the security door,
// then go back to the door, and work out which items we need to hold to let us through.
// Some items are traps: when we realise we have fallen into a trap, restart and remember
// not to pick up that item next time.

// Either find a trap, or end up where we started with all non-trap items,
// and (hopefully) knowing where the security door is (a stack of moves)
fn parta_traverse<T: 'static, C: 'static>(
    traps: &HashSet<String>,
    ret: Option<Door>,
    mut game: Game<T, C>,
) -> Result<(Game<T, C>, Option<Vec<Door>>), String>
where
    C: Vector<T>,
    T: Clone,
    T: TryInto<usize>,
    <T as TryInto<usize>>::Error: Debug,
    T: ToPrimitive,
    T: From<u8>,
    T: Num,
    T: PartialOrd,
    T: Display,
{
    let mut sec_door = None;
    for i in game.room.items.clone() {
        if !traps.contains(&i) {
            game.do_action(Cmd::Take(i))?;
        }
    }
    for d in game.room.doors.clone() {
        if ret.map_or(false, |r| r == d) {
            continue;
        }
        if game.do_action(Cmd::Move(d.clone()))? {
            // found security door
            sec_door = Some(vec![d]);
        } else {
            let tmp = parta_traverse(traps, Some(rev(&d)), game)?;
            game = tmp.0;
            let sec_door_tmp = tmp.1;
            match sec_door_tmp {
                None => {}
                Some(ms) => {
                    let mut tmp = ms;
                    tmp.push(d);
                    sec_door = Some(tmp)
                }
            }
        }
    }
    match ret {
        None => {}
        Some(r) => {
            game.do_action(Cmd::Move(r))?;
        }
    }
    Ok((game, sec_door))
}

#[derive(Clone, Copy, Debug)]
enum Gray<T> {
    Up(T),
    Down(T),
}
impl<T> Gray<T> {
    fn flip(self) -> Self {
        match self {
            Gray::Up(x) => Gray::Down(x),
            Gray::Down(x) => Gray::Up(x),
        }
    }
}
// A Gray code on items.len() bits, expressed as transitions of named bits
// assuming we start from all bits 0
fn gray<T: Clone>(items: &[T]) -> impl Iterator<Item = Gray<&T>> {
    fn gray1<T: Clone>(items: &[T]) -> Vec<Gray<&T>> {
        if items.len() == 0 {
            return Vec::new();
        }
        let rest = gray1(&items[1..items.len()]);
        let mut ret: Vec<Gray<&T>> = Vec::with_capacity(rest.len() * 2 + 1);
        ret.extend(rest.clone());
        ret.push(Gray::Up(&items[0]));
        ret.extend(rest.into_iter().rev().map(Gray::flip));
        ret
    }
    gray1(items).into_iter()
}

fn parta<T: 'static, C: 'static>(mach: IC<T, C>) -> String
where
    C: Vector<T>,
    T: Clone,
    C: Clone,
    T: TryInto<usize>,
    <T as TryInto<usize>>::Error: Debug,
    T: ToPrimitive,
    T: From<u8>,
    T: Num,
    T: PartialOrd,
    T: Display,
{
    fn mk_game<T: 'static, C: 'static>(mach: IC<T, C>) -> Game<T, C>
    where
        C: Vector<T>,
        T: Clone,
        T: TryInto<usize>,
        <T as TryInto<usize>>::Error: Debug,
        T: ToPrimitive,
        T: From<u8>,
        T: Num,
        T: PartialOrd,
        T: Display,
    {
        let (room1, kont) = run_to_in(mach);
        let room = parse_room(&room1).unwrap();
        match room {
            ParsedRoom::Room(room) => Game {
                mach: kont,
                room,
                inv: Vec::new(),
                last_pickup: None,
            },
            ParsedRoom::EjectedFromCheckpoint(_) => panic!("Start should not be in the checkpoint"),
        }
    }

    let mut traps = HashSet::new();

    let (mut game, mut sec_door) = loop {
        match parta_traverse(&traps, None, mk_game(mach.clone())) {
            Err(trap) => {
                println!("Found trap: {}", trap);
                traps.insert(trap);
                continue;
            }
            Ok((_, None)) => panic!("Traversed and not found security checkpoint"),
            Ok((game, Some(sec_door))) => {
                break (game, sec_door);
            }
        }
    };
    println!(
        "Got all items, now seaching for correct set to pass security checkpoint: {:?}",
        game.inv
    );
    println!("The security door is at {:?}", sec_door);
    while sec_door.len() > 1 {
        let d = sec_door.pop().unwrap();
        game.do_action(Cmd::Move(d)).unwrap();
    }
    let security = sec_door.pop().unwrap();

    // Find the correct set of items to get through the checkpoint
    // We could be more efficient, because we get told "heavier" or "lighter", but brute force
    // is fast enough. Using Gray code means we only pick up/drop one item between each test.
    let all_items = game.inv.clone();
    let mut gray_items = gray(&all_items);
    game.do_action(Cmd::Move(security)).unwrap();
    while game.room.name != "Pressure-Sensitive Floor" {
        match gray_items.next() {
            Some(Gray::Up(i)) => game.do_action(Cmd::Drop(i.to_string())).unwrap(),
            Some(Gray::Down(i)) => game.do_action(Cmd::Take(i.to_string())).unwrap(),
            None => panic!("Could not get through door using items: {:?}!", all_items),
        };
        game.do_action(Cmd::Move(security)).unwrap();
    }
    game.room.desc.lines().nth(2).unwrap().to_string()
}

fn main() {
    let mem: Sparse<usize, i64> = read_intcode_sparse("../data/day25");
    let mach = IC::new(0, mem.clone());

    println!("part a: {}", parta(mach));
}

File: 2019/haskell/Day25.hs
{-# LANGUAGE LambdaCase, TemplateHaskell #-}
module Main where

import Control.Applicative ((<|>), many)
import Control.Lens (makeLenses, (^.),(&),(.~),(%~),(^?!),_Just)
import Control.Monad.State (State,get,put,evalState,state)
import Data.Bifunctor (first)
import Data.Char (chr,ord)
import Data.Foldable (asum, traverse_)
import Data.Functor ((<&>))
import Data.Maybe (fromJust)
import qualified Data.Set as S
import Data.Traversable (for)
import Data.Void(Void)
import Text.Megaparsec (Parsec, ParseErrorBundle, errorBundlePretty, runParser, eof, manyTill, try)
import Text.Megaparsec.Char (eol, printChar, string)

import Intcode

-- Line-based, string-based interface to the droid.
-- The function will newline-terminate the input.
-- We panic if the IC doesn't accept the whole line.
-- The maybe tells us when the machine halts.
runToIn' :: IC -> (String, Maybe (String -> IC))
runToIn' ic = case runToIO ic of
  ICOut c ic' -> first (chr (fromInteger c):) $ runToIn' ic'
  ICIn f -> ("", Just $ giveLine f)
  ICHalt _ -> ("", Nothing)
  where giveLine :: (Integer -> IC) -> String -> IC
        giveLine f "" = f $ toInteger $ ord '\n'
        giveLine f (c:cs) =  case runToIO $ f $ toInteger $ ord c of
          ICIn f' -> giveLine f' cs
          _ -> error "giveLine: machine did not want all the input"

newtype StringIC = SIC {runSIC :: String -> (String, Maybe StringIC)}
runToIn :: IC -> (String, Maybe StringIC)
runToIn ic = let (s,x) = runToIn' ic
             in (s,fmap (SIC . (.) runToIn) x)

type Item = String

data Door = N | E | S | W deriving (Eq, Show)

reverseDoor :: Door -> Door
reverseDoor N = S
reverseDoor E = W
reverseDoor S = N
reverseDoor W = E

data Room = Room {_name :: String
                 ,_desc :: String
                 ,_doors :: [Door]
                 ,_items :: [Item]
                 }
  deriving Show
$(makeLenses ''Room)

data Weight = TooHeavy | TooLight
data RejectRoom = AcceptRoom Room -- this is a actual room which we have parsed
                | RejectRoom Weight Room -- hit security checkpoint, and got kicked out, the 'Room' is the one we were kicked into

parseRoom :: String -> Either (ParseErrorBundle String Void) RejectRoom
parseRoom = runParser parseRoom' ""

parseRoom' :: Parsec Void String RejectRoom
parseRoom' = try (AcceptRoom <$> oneRoom)
         <|> uncurry RejectRoom <$> securityThenOneRoom
  where oneRoom = mkRoom <$> nameP <*> descP <*> doorsP <*> (Left <$> itemsP <* promptP <|> Right <$> success) <* eof
        mkRoom n des dor (Left item) = Room n des dor item
        mkRoom n des dor (Right santa) = Room n (unlines [des,santa]) dor []
        securityThenOneRoom = (,) <$ nameP <* descP <* doorsP <*> securityP <*> oneRoom <* eof
        nameP = eol *> eol *> eol *> string "== " *> manyTill printChar (string " ==") <* eol
        descP = manyTill printChar eol <* eol
        doorsP = string "Doors here lead:" *> eol *> parseList directionP <* eol
        directionP = N <$ string "north"
                 <|> E <$ string "east"
                 <|> S <$ string "south"
                 <|> W <$ string "west"
        itemsP = string "Items here:" *> eol *> parseList (many printChar) <* eol
             <|> pure []
        promptP = string "Command?" <* eol
        success = (\l1 l2 l3 -> unlines [l1,l2,l3])
              <$> string "A loud, robotic voice says \"Analysis complete! You may proceed.\" and you enter the cockpit." <* eol
              <*> line
              <*> line
        line = many printChar <* eol
        securityP = string "A loud, robotic voice says \"Alert! Droids on this ship are "
                 *> (TooLight <$ string "heavier" <|> TooHeavy <$ string "lighter")
                 <* string " than the detected value!\" and you are ejected back to the checkpoint."
                 <* eol

parseList :: Parsec Void String p -> Parsec Void String [p]
parseList p = many (string "- " *> p <* eol)

parseInventory :: Parsec Void String [Item]
parseInventory = eol
              *> string "Items in your inventory:"
              *> eol
              *> parseList (many printChar)
              <* eol
              <* string "Command?"
              <* eol
              <* eof


data Game = Game {_traps :: [Item] -- we don't actually need to store this, but it is nice to report traps at the end
                 ,_beforeLastPickup :: Maybe (Game, Item)
                 ,_movesSinceLastPickup :: [Door] -- latest first
                 ,_droid :: StringIC
                 }
$(makeLenses ''Game)

data MoveResult = Ok Game -- success (includes bouncing off the security checkpoint)
                | Trap Game -- noticed last pickup was a trap, have rewound to not pick it up, and then replayed moves
                            -- i.e. success, but have lost an item

-- Returns Nothing if the move failed (i.e. we didn't seem to reach a room - probably we have picked up a trap!)
move' :: Door -> Game -> (MoveResult, RejectRoom)
move' d g = let (r,d') = (g ^. droid) `runSIC` doorString d
                room' = parseRoom r
            in case (room',d') of
                 (Right room@(AcceptRoom _), Just newDroid) -> (Ok $ g & droid .~ newDroid & movesSinceLastPickup %~ (d:), room)
                 (Right room@(RejectRoom _ _), Just newDroid) -> (Ok $ g & droid .~ newDroid, room)
                 (Right room, Nothing) -> (Ok undefined, room) -- This is rather unsafe. It should only happen when we have just gotten
                                                               -- through the security door, and then will never inspect the undefined.
                 _ -> case move' d $ untrap g of -- assume last picked up item was trap
                   (Ok g', r') -> (Trap g', r')
                   (Trap _, _) -> error "Uh oh, multiple traps?" -- If I understand the puzzle correctly, this should not happen!
  where doorString N = "north"
        doorString E = "east"
        doorString S = "south"
        doorString W = "west"
        move'' :: Door -> Game -> Game
        -- errors if the move fails, for use when replaying a series
        -- of moves after un-picking the trap (so we believe this should never fail)
        move'' door game = case move' door game of
          (Ok g', _) -> g'
          _ -> error "Replaying moves failed - our trap detection is buggy!"
        untrap :: Game -> Game
        -- Rewind and replay so the last item was never picked up
        untrap game = let (prior, trap) = game ^?! beforeLastPickup._Just
                          since = game ^. movesSinceLastPickup
                          new = foldr move'' prior since
                      in new & traps %~ (trap:)

-- Do the move, silently dropping old pickups if necessary
move :: Door -> State Game RejectRoom
move d = state $ \g -> case move' d g of
  (Ok g', r) -> (r, g')
  (Trap g', r) -> (r, g')

-- Unfortunately, "apply f, but if it fails then backtrack and retry" doesn't form the bind of a monad
-- so we can't get nice do notation for backtracking. Instead we will do this step manually, using move.

pickup :: Item -> State Game ()
pickup i = do g <- get
              let (r,d') = (g ^. droid) `runSIC` ("take " ++ i)
              let expected = unlines [""
                                     ,"You take the " ++ i ++ "."
                                     ,""
                                     ,"Command?"
                                     ]
              let newStateIfTrap = g & traps %~ (i:)
              if r /= expected
                then put newStateIfTrap
                else case d' of
                       Nothing -> put newStateIfTrap
                       Just droid' -> put $ Game {_traps = g ^. traps
                                                 ,_beforeLastPickup = Just (g, i)
                                                 ,_movesSinceLastPickup = []
                                                 ,_droid = droid'
                                                 }

putdown :: Item -> State Game ()
putdown i = do g <- get
               let (r,d') = (g ^. droid) `runSIC` ("drop " ++ i)
               let expected = unlines [""
                                      ,"You drop the " ++ i ++ "."
                                      ,""
                                      ,"Command?"
                                      ]
               if r /= expected
                 then error "Didn't expect a trap upon dropping"
                 else case d' of
                        Nothing -> error "Didn't expect a trap upon dropping"
                        Just droid' -> put $ g & droid .~ droid'

inventory :: State Game [Item]
inventory = do g <- get
               let (inv', _) = (g ^. droid) `runSIC` "inv"
               case runParser parseInventory "" inv' of
                 Left err -> error (errorBundlePretty err)
                 Right inv -> pure inv

data Transition a = Up a | Down a

-- Gray code expressed as named bit transitions
grayPicks :: [a] -> [Transition a]
grayPicks [] = []
grayPicks (b:bs) = let lower = grayPicks bs
                   in lower ++ [Up b] ++ map flipT (reverse lower)
  where flipT (Up a) = Down a
        flipT (Down a) = Up a

-- Assume the ship is a tree with backedges.
-- Do a preorder traversal of the ship picking up all items, detecting the traps and recording where the exit is.
-- Then go to the exit and search for the set of items that will let us pass.
-- Return whatever text the game prints after passing the exit
parta :: Room -> State Game String
parta r = pickupAll Nothing r >>= \case
            Nothing -> error "Couldn't find security checkpoint"
            Just (ds, d) -> walk ds >> passDoor d
  where pickupAll :: Maybe Door -> Room -> State Game (Maybe ([Door],Door))
        -- preorder traversal; returns the path to the room before the pressure pad and the door to the pad
        pickupAll ent room = do case room ^. items of
                                  [] -> pure ()
                                  [i] -> pickup i
                                  is -> error $ "Did not expect multiple items in a room! Found: " ++ unwords is
                                  -- because then how would we tell which was a trap when we next move?
                                let ret = reverseDoor <$> ent
                                childrenPath <- for (filter ((/=ret).Just) $ room ^. doors) $ \d ->
                                  move d >>= \case AcceptRoom r' -> pickupAll (Just d) r'
                                                   RejectRoom _ _ -> pure (Just ([],d))
                                let path' = asum childrenPath
                                case ent of
                                  Nothing -> pure path'
                                  Just ent' -> move (reverseDoor ent') >> pure (first (ent':) <$> path')
        walk :: [Door] -> State Game ()
        walk = traverse_ move
        passDoor :: Door -> State Game String
        -- We want to drop some subset of our inventory to be the correct weight to get through security
        passDoor d = do inv <- S.fromList <$> inventory
                        go (grayPicks $ S.toList inv) inv S.empty S.empty S.empty
          where go :: [Transition Item] -> S.Set Item -> S.Set Item -> S.Set (S.Set Item) -> S.Set (S.Set Item) -> State Game String
                -- Iterate through subsets using Gray code. Skip testing sets whose answer is determined by
                -- prior knowledge.
                go [] _ _ _ _ = error "passDoor: ran out of options"
                go (p:ps) inv flr tooLight tooHeavy =
                  do tooLightHeavy' <- if inv `S.member` tooLight || flr `S.member` tooHeavy
                                       then pure $ Left (tooLight, tooHeavy)
                                       else move d <&> \case
                                         AcceptRoom r' -> Right $ r' ^. desc
                                         RejectRoom w _ -> Left $ case w of
                                           TooLight -> (tooLight `S.union` S.powerSet inv, tooHeavy)
                                           TooHeavy -> (tooLight, tooHeavy `S.union` S.powerSet flr)
                     case tooLightHeavy' of
                       Right s -> pure s
                       Left (tooLight', tooHeavy') -> case p of
                         Down i -> pickup i *> go ps (S.insert i inv) (S.delete i flr) tooLight' tooHeavy'
                         Up i -> putdown i *> go ps (S.delete i inv) (S.insert i flr) tooLight' tooHeavy'

mkGame :: IC -> Either (ParseErrorBundle String Void) (Room, Game)
mkGame mach = let (r,f) = runToIn mach
                  g = Game {_traps = [], _beforeLastPickup = Nothing, _movesSinceLastPickup = [], _droid = fromJust f}
              in do r' <- parseRoom r
                    case r' of
                      AcceptRoom r'' -> pure (r'', g)
                      RejectRoom _ _ -> error "mkGame: first room can't be checkpoint!"

main :: IO ()
main = do mach <- readIntcode "../data/day25"
          case mkGame mach of
            Left err -> putStr $ errorBundlePretty err
            Right (r,g) -> putStrLn $ evalState (parta r) g
